Mathlib/Algebra/MonoidAlgebra/Equiv.lean

```diff
@@ -0,0 +1,143 @@
+      classical
+      refine Finsupp.ext fun a => ?_
+      simp [one_def]
+    map_mul' := fun x y => by
+      classical
+      -- Porting note: was `ext`
+      refine Finsupp.ext fun a => ?_
+      simp_rw [mul_apply, mul_def, Finsupp.mapRange_apply]
+      rw [Finsupp.sum_apply, map_finsupp_sum f]
+      rw [Finsupp.sum_mapRange_index];
+      -- Porting note: was `congrm`
+      apply congr_arg _ <| funext₂ fun g1 r1 => ?_
+      rw [Finsupp.sum_mapRange_index];
+      rw [Finsupp.sum_apply, map_finsupp_sum f]
+      -- Porting note: was `congrm`
+      apply congr_arg _ <| funext₂ fun g2 r2 => ?_
+      · rw [Finsupp.single_apply]
+        split_ifs with h <;> simp only [h, if_false, if_true, map_mul, map_zero]
+      all_goals
+        intro; simp only [MulZeroClass.mul_zero, MulZeroClass.zero_mul];
+        simp only [ite_self, Finsupp.sum_zero] }
+
+-- TODO: times out with `AddZeroClass G`
+@[simps!]
+def ringEquivCongrLeft {R S G : Type _} [Semiring R] [Semiring S] [AddMonoid G] (f : R ≃+* S) :
+    AddMonoidAlgebra R G ≃+* AddMonoidAlgebra S G :=
+  .ofHomInv (ringHomCongrLeft f) (ringHomCongrLeft f.symm) (by ext <;> simp) (by ext <;> simp)
+
+@[simps]
+def algHomCongrLeft {k R S G : Type _} [CommSemiring k] [Semiring R] [Algebra k R] [Semiring S]
+    [Algebra k S] [AddMonoid G] (f : R →ₐ[k] S) : AddMonoidAlgebra R G →ₐ[k] AddMonoidAlgebra S G :=
+  { ringHomCongrLeft f.toRingHom with
+    toFun := (Finsupp.mapRange f f.map_zero : AddMonoidAlgebra R G → AddMonoidAlgebra S G)
+    commutes' := fun r => by
+      -- Porting note: was `ext`
+      refine Finsupp.ext fun a => ?_
+      simp_rw [AddMonoidAlgebra.coe_algebraMap, Function.comp_apply, Finsupp.mapRange_single]
+      congr 2
+      rw [AlgHom.map_algebraMap] }
+
+@[simps!]
+def algEquivCongrLeft {k R S G : Type _} [CommSemiring k] [Semiring R] [Algebra k R] [Semiring S]
+    [Algebra k S] [AddMonoid G] (f : R ≃ₐ[k] S) : AddMonoidAlgebra R G ≃ₐ[k] AddMonoidAlgebra S G :=
+  .ofAlgHom (algHomCongrLeft f) (algHomCongrLeft f.symm)
+    (AlgHom.ext fun x => Finsupp.ext fun a => by simp)
+    (AlgHom.ext fun x => Finsupp.ext fun a => by simp)
+
+@[simps]
+def algAutCongrLeft {k R G : Type _} [CommSemiring k] [Semiring R] [Algebra k R] [AddMonoid G] :
+    (R ≃ₐ[k] R) →* AddMonoidAlgebra R G ≃ₐ[k] AddMonoidAlgebra R G where
+  toFun f := algEquivCongrLeft f
+  map_one' := by
+    ext
+    refine Finsupp.ext fun a => ?_
+    simp [Finsupp.mapRange_apply]
+  map_mul' x y := by
+    ext
+    refine Finsupp.ext fun a => ?_
+    simp [Finsupp.mapRange_apply]
+
+@[simp 1001] -- LHS simplifies
+lemma algAutCongrLeft_apply_one {k R G : Type _} [CommSemiring k] [Semiring R] [Algebra k R]
+    [AddMonoid G] :
+    algAutCongrLeft (k := k) (R := R) (G := G) 1 = AlgEquiv.refl := by
+  ext
+  exact Finsupp.mapRange_id _
+
+@[simps]
+def mapDomainRingEquiv (k : Type _) [Semiring k] {G H : Type _} [AddMonoid G] [AddMonoid H]
+    (f : G ≃+ H) : AddMonoidAlgebra k G ≃+* AddMonoidAlgebra k H :=
```

Comment:
This falls out trivially from `domCongr` with `k=Nat`, though perhaps that's not a satisfying way to get it.

Reply:
Indeed! I've forgotten we can take `k := Nat` rather than `k`, even though I've [used this trick before](https://github.com/leanprover-community/mathlib/pull/17401/files#diff-a97e4b387e41d31d3f5467f44ef6cf00d77f7278449af9d55821b61a2f4affc0L1345-R1340) to generalize to the noncommutative setting. Now I'm confused: what's not satisfying about it?

Reply:
Usually we build stronger bundlings from weaker bundlings (as in general this scales), rather than building the strongest one first and recovering the weaker one as a special case.

I suspect though that no one needs the weaker one in the first place...